apollo-federation/src/merger/merger.rs

```diff
@@ -1197,12 +1213,143 @@ impl Merger {
+        let mut joins_by_directive_name: HashMap<
+            Name,
+            HashMap<Vec<Node<Argument>>, IndexSet<Name>>,
+        > = HashMap::new();
+        let mut links_to_persist: Vec<(Url, Directive)> = Vec::new();
+
+        for (idx, source) in sources.iter() {
+            let Some(source) = source else {
+                continue;
+            };
+            let graph = self.join_spec_name(*idx)?;
+            let schema = self.subgraphs[*idx].schema();
+            let Some(link_import_identity_url_map) = schema.metadata() else {
+                continue;
+            };
+            let Ok(Some(link_directive_name)) = self
+                .link_spec_definition
+                .directive_name_in_schema(schema, &DEFAULT_LINK_NAME)
+            else {
+                continue;
+            };
+
+            let source: DirectiveTargetPosition = source.clone().try_into()?;
+            for directive in source.get_all_applied_directives(schema).iter() {
+                let mut should_include_as_join_directive = false;
+
+                if directive.name == link_directive_name {
+                    if let Ok(link) = Link::from_directive_application(directive) {
+                        should_include_as_join_directive =
+                            self.should_use_join_directive_for_url(&link.url);
+
+                        if should_include_as_join_directive
+                            && SPEC_REGISTRY.get_definition(&link.url).is_some()
+                        {
+                            links_to_persist.push((link.url.clone(), directive.as_ref().clone()));
+                        }
+                    }
+                } else if let Some(url_for_directive) =
+                    link_import_identity_url_map.source_link_of_directive(&directive.name)
+                {
+                    should_include_as_join_directive =
+                        self.should_use_join_directive_for_url(&url_for_directive.link.url);
+                }
+
+                if should_include_as_join_directive {
+                    let existing_joins = joins_by_directive_name
+                        .entry(directive.name.clone())
+                        .or_default();
+                    let existing_graphs_with_these_arguments = existing_joins
+                        .entry(directive.arguments.clone())
+                        .or_default();
+                    existing_graphs_with_these_arguments.insert(graph.clone());
+                }
+            }
+        }
+
+        let Some(link_directive_name) = self
+            .link_spec_definition
+            .directive_name_in_schema(&self.merged, &DEFAULT_LINK_NAME)?
+        else {
+            bail!(
+                "Link directive must exist in the supergraph schema in order to apply join directives"
+            );
+        };
+
+        let mut latest_or_highest_link_by_identity: HashMap<Identity, (Url, Directive)> =
```

Comment:
can you add the comment from [merge.ts#L3214-L3223](https://github.com/apollographql/federation/blob/main/composition-js/src/merging/merge.ts#L3214-L3223) that explains why we do this?